File: rocky/capture/session.py
"""
Session tracking for Rocky.

Enforces two smart silence rules:
  - Daily quiz budget: max 3 Socratic loops per day (configurable)
  - Cool-down window: no quiz within 2 hours of the last one (configurable)

Also logs every task description for future prompt quality tracking (Phase 5).
State is stored in the main PKG database so nothing extra to manage.
"""

import sqlite3
from datetime import date, datetime, timedelta
from pathlib import Path

from rocky.graph.store import DB_FILE

_DEFAULT_DAILY_BUDGET = 3
_DEFAULT_MIN_GAP_MINUTES = 120


class Session:
    def __init__(self, db_path: Path = DB_FILE,
                 daily_budget: int = _DEFAULT_DAILY_BUDGET,
                 min_gap_minutes: int = _DEFAULT_MIN_GAP_MINUTES):
        self.daily_budget = daily_budget
        self.min_gap_minutes = min_gap_minutes
        self._db = db_path

    def _connect(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self._db)
        conn.row_factory = sqlite3.Row
        return conn

    def _get(self, key: str) -> str | None:
        with self._connect() as conn:
            row = conn.execute(
                "SELECT value FROM session_state WHERE key = ?", (key,)
            ).fetchone()
            return row["value"] if row else None

    def _set(self, key: str, value: str):
        now = datetime.now().isoformat()
        with self._connect() as conn:
            conn.execute("""
                INSERT INTO session_state (key, value, updated_at)
                VALUES (?, ?, ?)
                ON CONFLICT(key) DO UPDATE SET value = excluded.value, updated_at = excluded.updated_at
            """, (key, value, now))

    # ── budget ────────────────────────────────────────────────────────────────

    def budget_remaining(self) -> int:
        """Quizzes remaining today. Resets automatically at midnight."""
        last_reset = self._get("budget_reset_date")
        today = date.today().isoformat()

        if last_reset != today:
            self._set("budget_reset_date", today)
            self._set("budget_used", "0")
            return DAILY_BUDGET

        used = int(self._get("budget_used") or "0")
        return max(0, self.daily_budget - used)

    def record_quiz(self):
        """Call once per completed Socratic loop to decrement the budget."""
        used = int(self._get("budget_used") or "0")
        self._set("budget_used", str(used + 1))
        self._set("last_quiz_at", datetime.now().isoformat())

    # ── cool-down ─────────────────────────────────────────────────────────────

    def cool_down_active(self) -> bool:
        """True if a quiz happened within the last MIN_GAP_MINUTES."""
        last = self._get("last_quiz_at")
        if not last:
            return False
        elapsed = datetime.now() - datetime.fromisoformat(last)
        return elapsed < timedelta(minutes=self.min_gap_minutes)

    def minutes_until_ready(self) -> int:
        last = self._get("last_quiz_at")
        if not last:
            return 0
        elapsed = datetime.now() - datetime.fromisoformat(last)
        remaining = timedelta(minutes=self.min_gap_minutes) - elapsed
        return max(0, int(remaining.total_seconds() / 60))

    # ── combined gate ─────────────────────────────────────────────────────────

    def can_quiz(self) -> tuple[bool, str]:
        """
        Returns (allowed, reason_if_blocked).
        Callers use this to decide whether to run the Socratic loop.
        """
        if self.cool_down_active():
            mins = self.minutes_until_ready()
            return False, f"cool-down active — Rocky ready again in ~{mins}m"
        if self.budget_remaining() == 0:
            return False, "daily quiz budget reached (3/day) — resets tomorrow"
        return True, ""

    # ── task logging ──────────────────────────────────────────────────────────

    def log_task(self, task: str, mode: str = "manual"):
        """Log a task description. Used by Phase 5 prompt quality tracking."""
        now = datetime.now().isoformat()
        with self._connect() as conn:
            conn.execute(
                "INSERT INTO task_log (task, mode, logged_at) VALUES (?, ?, ?)",
                (task, mode, now),
            )

File: src/session.rs
/// Session tracking — daily quiz budget and cool-down window.
use anyhow::Result;
use chrono::{Duration, Local};

use crate::db::Db;

pub struct Session {
    db: Db,
    pub daily_budget: u32,
    pub min_gap_minutes: u32,
}

impl Session {
    pub fn new(db: Db, daily_budget: u32, min_gap_minutes: u32) -> Self {
        Self { db, daily_budget, min_gap_minutes }
    }

    pub fn budget_remaining(&self) -> Result<u32> {
        let today = Local::now().date_naive().to_string();
        let last_reset = self.db.session_get("budget_reset_date")?;
        if last_reset.as_deref() != Some(&today) {
            self.db.session_set("budget_reset_date", &today)?;
            self.db.session_set("budget_used", "0")?;
            return Ok(self.daily_budget);
        }
        let used: u32 = self.db.session_get("budget_used")?
            .unwrap_or_default()
            .parse()
            .unwrap_or(0);
        Ok(self.daily_budget.saturating_sub(used))
    }

    pub fn record_quiz(&self) -> Result<()> {
        let used: u32 = self.db.session_get("budget_used")?
            .unwrap_or_default()
            .parse()
            .unwrap_or(0);
        self.db.session_set("budget_used", &(used + 1).to_string())?;
        self.db.session_set("last_quiz_at", &Local::now().naive_local().to_string())?;
        self.db.increment_total_quizzes()?;
        Ok(())
    }

    pub fn cool_down_active(&self) -> Result<bool> {
        let last = self.db.session_get("last_quiz_at")?;
        if let Some(last) = last {
            if let Ok(dt) = chrono::NaiveDateTime::parse_from_str(&last, "%Y-%m-%d %H:%M:%S%.f") {
                let elapsed = Local::now().naive_local() - dt;
                return Ok(elapsed < Duration::minutes(self.min_gap_minutes as i64));
            }
        }
        Ok(false)
    }

    pub fn minutes_until_ready(&self) -> Result<i64> {
        let last = self.db.session_get("last_quiz_at")?;
        if let Some(last) = last {
            if let Ok(dt) = chrono::NaiveDateTime::parse_from_str(&last, "%Y-%m-%d %H:%M:%S%.f") {
                let elapsed = Local::now().naive_local() - dt;
                let remaining = Duration::minutes(self.min_gap_minutes as i64) - elapsed;
                return Ok(remaining.num_minutes().max(0));
            }
        }
        Ok(0)
    }

    /// Returns (allowed, reason_if_blocked). Enforces both cooldown and daily budget.
    /// Use for automatic triggers (git hooks, Claude Code hook).
    pub fn can_quiz(&self) -> Result<(bool, String)> {
        if self.cool_down_active()? {
            let mins = self.minutes_until_ready()?;
            return Ok((false, format!("cool-down active — Rocky ready again in ~{mins}m")));
        }
        if self.budget_remaining()? == 0 {
            return Ok((false, format!("daily quiz budget reached ({}/day) — resets tomorrow", self.daily_budget)));
        }
        Ok((true, String::new()))
    }

    /// Always allows quizzing — no cooldown, no budget cap.
    /// Use when the user explicitly invokes Rocky, since they're choosing to learn.
    pub fn can_quiz_manual(&self) -> Result<(bool, String)> {
        Ok((true, String::new()))
    }

    /// Call once at the start of a quiz session to update the streak counter.
    /// Returns the new streak value.
    pub fn update_streak(&self) -> Result<u32> {
        let today = Local::now().date_naive().to_string();
        let yesterday = (Local::now().date_naive() - Duration::days(1)).to_string();
        let last_date = self.db.session_get("last_quiz_date")?;
        let current: u32 = self.db.session_get("quiz_streak")?
            .unwrap_or_default().parse().unwrap_or(0);
        let new_streak = match last_date.as_deref() {
            Some(d) if d == today     => current,       // already counted today
            Some(d) if d == yesterday => current + 1,   // consecutive day
            _                         => 1,             // first time or broken
        };
        self.db.session_set("last_quiz_date", &today)?;
        self.db.session_set("quiz_streak", &new_streak.to_string())?;
        Ok(new_streak)
    }

}
